libp2p/Service.cpp

```diff
@@ -1193,20 +1195,30 @@ bool Service::updatePeersToIni(std::map<dev::network::NodeIPEndpoint, NodeID> co
         auto pos1 = fileData.find("p2p");
         if (pos1 == fileData.npos)
         {
+            SERVICE_LOG(DEBUG) << LOG_DESC("fail to locate peers' information in config file");
             return false;
         }
         auto pos2 = fileData.find("node.0=", pos1);
         auto pos3 = fileData.find("certificate_blacklist");
         if (pos2 == fileData.npos || pos3 == fileData.npos || pos2 >= pos3)
         {
+            SERVICE_LOG(DEBUG) << LOG_DESC("fail to locate peers' information in config file");
             return false;
         }
         fileData.replace(pos2, pos3 - pos2 - 1, tmpdata);
 
-        // TODO: 写入保护
-        std::ofstream out(confdir, std::ios::out);
-        out << fileData;
-        out.close();
+        auto newFileDir = confdir + ".new";
+        std::ofstream outfile(newFileDir, std::ios::out);
+        outfile.flush();
+        outfile << fileData;
+        outfile.close();
+
+        if (0 != rename(newFileDir.c_str(), confdir.c_str()))
```

Comment:
if the rename operation failed, it means that the old config.ini maybe remove and the exist config is the new config.ini, and when the node restart, you should check this situation.

Reply:
[https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html](url)says “If the rename() function fails for any reason other than [EIO], any file named by new shall be unaffected.”

Reply:
i wonder if it is save enough to use rename